doc/adr/2019-01-04-error-reporting-with-statusor.md

```diff
@@ -0,0 +1,39 @@
+**Title**: Functions should report errors using `StatusOr<T>` *instead of*
+throwing exceptions.
+
+**Status**: proposed
+
+**Context**: We know there will be users of these C++ libraries who want to use
+C++ exceptions as well as those who are not able to. Our C++ libraries must
+work for all of our users, regardless of their ability to use exceptions.
+
+**Decision**: All of our APIs will report errors to callers by returning a
```

Comment:
"All of our APIs" may be tricky: what about things that return `void`?  Are we going to return `StatusOr<void>` or `Status`?  What about returning an object derived from `std::istream` (or `std::ostream`) which already have a mechanism to report errors (the `badbit`)?

Consider starting with "None of our APIs will throw exceptions..." (modulo platforms that raise `std::bad_alloc` but I digress), and then saying "Our APIs will return `StatusOr<T>` <placeholder> to indicate an error. The only exception would be APIs that have existing error reporting mechanisms, e.g., `std::ostream`"

In `<placeholder>` you may want to say "(except APIs that return `void` in which case the API will return `Status`)" or something like that.

Reply:
I forgot to mention, if we return `Status` instead of `StatusOr<void>` there is no `.value()` to raise an exception from.

Reply:
All good. Incorporated.